cmd/juju/interact/query.go

```diff
@@ -13,36 +13,57 @@ import (
 // QueryVerify writes a question to w and waits for an answer to be read from
 // scanner.  It will pass the answer into the verify function.  Verify, if
 // non-nil, should check the answer for validity, returning an error that will
-// be written out to the user, or nil if answer is valid.
+// be written out to errOut, or nil if answer is valid.
 //
 // This function takes a scanner rather than an io.Reader to avoid the case
 // where the scanner reads past the delimiter and thus might lose data.  It is
 // expected that this method will be used repeatedly with the same scanner if
 // multiple queries are required.
-func QueryVerify(question []byte, scanner *bufio.Scanner, w io.Writer, verify func(string) error) (answer string, err error) {
-	defer fmt.Fprint(w, "\n")
+func QueryVerify(question string, scanner *bufio.Scanner, out, errOut io.Writer, verify VerifyFunc) (answer string, err error) {
+	defer fmt.Fprint(out, "\n")
 	for {
-		if _, err = w.Write(question); err != nil {
+		if _, err = out.Write([]byte(question)); err != nil {
 			return "", err
 		}
 
-		if !scanner.Scan() {
+		done := !scanner.Scan()
+
+		if done {
 			if err := scanner.Err(); err != nil {
 				return "", err
 			}
-			return "", io.EOF
 		}
 		answer = scanner.Text()
+		if done && answer == "" {
+			// EOF
+			return "", io.EOF
+		}
 		if verify == nil {
 			return answer, nil
 		}
-		err := verify(answer)
+		ok, msg, err := verify(answer)
+		if err != nil {
+			return "", err
+		}
 		// valid answer, return it!
-		if err == nil {
+		if ok {
 			return answer, nil
 		}
+
+		if done {
+			// can't query any more, nothing we can do.
```

Comment:
There might be a message.... I guess it's possible we got io.EOF from stdin but still can print to stdout. However, it's pretty likely that everything is going to have to error out at this point.  It should be one of those "it'll never happen" things, but we can try to handle it gracefully anyway.